sw/host/tests/chip/sysrst_ctrl/sysrst_ctrl_reset.rs

```diff
@@ -0,0 +1,181 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::{ensure, Result};
+use clap::Parser;
+use once_cell::sync::Lazy;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::time::Duration;
+
+use opentitanlib::app::TransportWrapper;
+use opentitanlib::io::gpio::PinMode;
+use opentitanlib::io::uart::Uart;
+use opentitanlib::test_utils::init::InitializeTest;
+use opentitanlib::test_utils::test_status::TestStatus;
+use opentitanlib::uart::console::UartConsole;
+use opentitanlib::{collection, execute_test};
+
+use sysrst_ctrl::{Config, setup_pins, set_pins};
+
+#[derive(Debug, Parser)]
+struct Opts {
+    #[command(flatten)]
+    init: InitializeTest,
+
+    /// Console receive timeout.
+    #[arg(long, value_parser = humantime::parse_duration, default_value = "10s")]
+    timeout: Duration,
+
+    /// Path to the firmware's ELF file, for querying symbol addresses.
+    #[arg(value_name = "FIRMWARE_ELF")]
+    firmware_elf: PathBuf,
+}
+
+static CONFIG: Lazy<HashMap<&'static str, Config>> = Lazy::new(|| {
+    collection! {
+        "hyper310" => Config {
+            // key0, key1, key2, pwrb_in, ac_present_in, lid_open
+            output_pins: vec!["IOR10", "IOR11", "IOR12", "IOR5", "IOR6", "IOR7"],
+            open_drain: vec![false, false, false, false, false, false],
+            input_pins: vec!["SYSRST_CTRL_EC_RST_L", "SYSRST_CTRL_FLASH_WP_L"],
+            //  ec_rst, flash_wp
+            pullup_pins: vec!["SYSRST_CTRL_EC_RST_L", "SYSRST_CTRL_FLASH_WP_L"],
+        },
+    }
+});
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+#[repr(u8)]
+#[allow(dead_code)]
+enum TestPhase {
+    CheckComboReset = 0,
+    CheckDeepSleepWakeup = 1,
+    CheckDeepSleepReset = 2,
+    FinalCheck = 3,
+}
+
+const PHASE_PINS: &[&str] = &["IOB0", "IOB1"];
+
+// Before CheckComboReset: all pins high.
+const PADS_VALUE_INIT: u8 = 0b111111;
+// Trigger reset in CheckComboReset: key0 and key1 low.
+// Also prepare for next phase: ac present low, power button high, lid open low.
+const PADS_VALUE_COMBO_RESET: u8 = 0b001100;
+// Trigger wakeup in CheckDeepSleepWakeup: lid open high.
+const PADS_VALUE_WAKEUP: u8 = 0b101100;
+// Trigger sleep reset in CheckDeepSleepReset: key2 and power button low.
+const PADS_VALUE_SLEEP_RESET: u8 = 0b100000;
+// Combo action detection time: set to around 5ms on the device.
+const COMBO_DETECTION_TIME_MICROS: u64 = 10_000;
+
+fn setup_phase_pins(transport: &TransportWrapper) -> Result<()> {
+    for pin in PHASE_PINS {
+        transport.gpio_pin(pin)?.set_mode(PinMode::PushPull)?;
+    }
+    Ok(())
+}
+
+fn set_test_phase(transport: &TransportWrapper, phase: TestPhase) -> Result<()> {
+    let mut val = phase as u8;
+    for pin in PHASE_PINS {
+        transport.gpio_pin(pin)?.write((val & 1) == 1)?;
+        val >>= 1;
+    }
+    ensure!(val == 0, "test phase value does not fit on the phase pins");
+    Ok(())
+}
+
+fn check_flash_wp_value(transport: &TransportWrapper, val: bool) -> Result<()> {
```

Comment:
Instead of a function, probably it's better to do
```rust
let flash_wp_pin = transport.gpio_pin("SYSRST_CTRL_FLASH_WP_L")?;
...
ensure!(flash_wp_pin().read()?, ...);
```